substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -468,130 +476,160 @@ where
 
 	/// Notify syncing state machine that a new sync peer has connected.
 	pub fn add_peer(&mut self, peer_id: PeerId, best_hash: B::Hash, best_number: NumberFor<B>) {
-		match self.add_peer_inner(peer_id, best_hash, best_number) {
-			Ok(Some(request)) =>
-				self.actions.push(ChainSyncAction::SendBlockRequest { peer_id, request }),
-			Ok(None) => {},
-			Err(bad_peer) => self.actions.push(ChainSyncAction::DropPeer(bad_peer)),
-		}
+		self.peers.insert(
+			peer_id,
+			PeerSync {
+				peer_id,
+				common_number: Zero::zero(),
+				best_hash,
+				best_number,
+				state: PeerSyncState::New,
```

Comment:
A new peer and an `Available` peer are handled differently in the code, and the idea was to postpone peer initialization that can lead to requests (`handle_new_peers()`) until `actions()`, so that we can call strategies in the specific order and implement the priorities of reserving the peers (e.g., Sync 2.0, then `ChainSync`, then `GapSync`).

Reply:
But `ChainSync::block_requests()` is also only called in `ChainSync::actions()` so is it not possible to insert the new peers as `PeerState::Available` and then deal with all peers the same way?

Reply:
I will need to look into this, as new peers are put into ancestry search state by default, and `ChainSync::block_requests()` deals with peers for which we already know the common number. May be we can simplify it, but it's not instantly obvious.